src/Columns/ColumnsCommon.h

```diff
@@ -70,6 +72,28 @@ ColumnPtr selectIndexImpl(const Column & column, const IColumn & indexes, size_t
                         ErrorCodes::LOGICAL_ERROR);
 }
 
+template <typename Column>
+size_t getLimitForPermutation(const Column & column, const IColumn::Permutation & perm, size_t limit)
+{
+    if (limit == 0)
+        limit = column.size();
+    else
+        limit = std::min(column.size(), limit);
+
+    if (perm.size() < limit)
+        throw Exception(ErrorCodes::SIZES_OF_COLUMNS_DOESNT_MATCH,
+            "Size of permutation ({}) is less than required ({})", perm.size(), limit);
```

Comment:
I'd prefer to pass `column.size()` and `perm.size()` into this function to be explicit about the fact that it operates only with sizes. 

But not sure that it is  matters.


```suggestion
inline size_t getLimitForPermutation(size_t column_size, size_t perm_size, size_t limit)
{
    if (limit == 0)
        limit = column_size;
    else
        limit = std::min(column_size, limit);

    if (perm_size < limit)
        throw Exception(ErrorCodes::SIZES_OF_COLUMNS_DOESNT_MATCH,
            "Size of permutation ({}) is less than required ({})", perm_size, limit);
```

Reply:
I agree, it's better.